### DISCUSSION TITLE
Metrics delivery reliability

### DISCUSSION THREAD

**Discussion by arnitolog** [grafana-cold-storage/agent#6190] [Q&A] [answered]
Hello,
I have Grafana Agent deployed in flow mode with controller.type: deployment. 
I did a small experiment and it looks like Agent can cover the case when Mimir is not reachable for a while. Does it use any local storage to keep metrics that are not deliverable or any kind of queue? If so, what are the default parameters, are they somehow configurable?

**Comment by tpaschalis** [comment, marked as answer]
Hey there @arnitolog. Yes, Grafana Agent reuses Prometheus' Write-Ahead Log (WAL) implementation to temporarily buffer data during network outages. More information about the data retention and parameters is available on [this](https://grafana.com/docs/agent/latest/flow/reference/components/prometheus.remote_write/#data-retention) page. Data is stored on the agent's `--storage.path` argument which defaults to a `data-agent` directory. You can influence the remote write behavior during both normal operations and remote_write outages using the `queue_config` and `wal` blocks, while defaults are documented on the component's [reference page](https://grafana.com/docs/agent/latest/flow/reference/components/prometheus.remote_write/)

Currently, the metrics WAL implementation does _not_ work for cases where the agent crashes or restarts; in these cases data _is_ lost but the WAL is only replayed to warm up a cache.

A similar WAL implementation also exists for logs in `loki.write`, primarily meant for network-based log sources.
